source/common/upstream/outlier_detection_impl.cc

```diff
@@ -282,29 +282,45 @@ DetectorImpl::~DetectorImpl() {
   }
 }
 
-std::shared_ptr<DetectorImpl> DetectorImpl::create(
-    const Cluster& cluster, const envoy::config::cluster::v3::OutlierDetection& config,
-    Event::Dispatcher& dispatcher, Runtime::Loader& runtime, TimeSource& time_source,
-    EventLoggerSharedPtr event_logger, Random::RandomGenerator& random) {
+std::shared_ptr<DetectorImpl>
+DetectorImpl::create(Cluster& cluster, const envoy::config::cluster::v3::OutlierDetection& config,
+                     Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
+                     TimeSource& time_source, EventLoggerSharedPtr event_logger,
+                     Random::RandomGenerator& random) {
   std::shared_ptr<DetectorImpl> detector(
       new DetectorImpl(cluster, config, dispatcher, runtime, time_source, event_logger, random));
 
   if (detector->config().maxEjectionTimeMs() < detector->config().baseEjectionTimeMs()) {
     throw EnvoyException(
         "outlier detector's max_ejection_time cannot be smaller than base_ejection_time");
   }
-
   detector->initialize(cluster);
 
   return detector;
 }
 
-void DetectorImpl::initialize(const Cluster& cluster) {
+void DetectorImpl::initialize(Cluster& cluster) {
   for (auto& host_set : cluster.prioritySet().hostSetsPerPriority()) {
     for (const HostSharedPtr& host : host_set->hosts()) {
       addHostMonitor(host);
     }
   }
+
+  if (cluster.healthChecker() != nullptr) {
+    cluster.healthChecker()->addHostCheckCompleteCb(
+        [this](HostSharedPtr host, HealthTransition changed_state) {
+          // If the host is ejected by outlier detection and active health check succeeds,
+          // we should treat this host as healthy.
+          if (Runtime::runtimeFeatureEnabled(
+                  "envoy.reloadable_features.successful_active_health_check_uneject_host") &&
+              changed_state == HealthTransition::Changed &&
```

Comment:
I don't think we want to gate on this? Since the health checker doesn't care about FAILED_OUTLIER_CHECK it would only set HealthTransition::Changed if we were already failing active hc and are now succeeding

I guess `!host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) && host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK)`  would be sufficient to say "we got a hc indicating that active hc is good, but the outlier detector has flagged this host so we should clear that out"? 

I think ideally we'd have an integration test operating between the active hc and the outlier detector, seems a bit tricky to make sure that things are working as expected just through unit tests

Reply:
>  I guess !host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) && host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK) would be sufficient to say "we got a hc indicating that active hc is good, but the outlier detector has flagged this host so we should clear that out"?

But we want this behaviour of whether active health check success should clear of pending outlier detection failure should be guarded by run time flag in case some one has some issues with it. But we are defaulting it to true.

Reply:
Sorry yeah this should be gated by a feature flag, I was thinking about the `HealthTransition::Changed` check. I would think that we want to respond to every health check result from the active health checker and reset any ejected hosts regardless of whether we transitioned the active health of the host?

Reply:
I see what you mean. Yes. That can be removed